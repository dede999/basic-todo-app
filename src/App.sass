#root
  h1
    text-align: center
    
  .container
    margin: 10vh
    display: flex
    flex-direction: column
    justify-content: space-between
    .content
      .title-box
        display: flex
        flex-direction: row
        justify-content: space-between
        .MuiButton-outlined
          margin-left: 15px
      .task-list
        display: flex
        flex-direction: column
        align-content: space-evenly
        margin-top: 5vh