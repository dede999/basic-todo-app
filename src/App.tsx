import './App.sass'
import React from 'react'
import TaskView from './components/TaskView'
import { Input, Button } from '@material-ui/core'
import ToDoContext, { ITask } from './providers/ToDoContext'

class App extends React.Component {

  state = {
    task_list: [] as ITask[],
    title: "Some task ..."
  }

  componentDidMount() {
    const tasks = localStorage.getItem("tasks")
    if(tasks) {
      this.setState({
        task_list: JSON.parse(tasks)
      })
    }
  }

  save_in_localStorage = () => {
    localStorage.setItem(
      "tasks", 
      JSON.stringify(this.state.task_list)
    )
  }
  
  add_task = () => {
    this.setState({
      task_list: [...this.state.task_list, {
        title: this.state.title,
        is_complete: false
      }],
      title: "Some Task ..."
    }, () => this.save_in_localStorage())
  }

  edit_task = (task: ITask, original: string) => {
    const index = this.state.task_list.findIndex(
      (arr_task: ITask) => original === arr_task.title
    )
    this.state.task_list.splice(index, 1, task)
    this.setState({
      task_list: this.state.task_list
    }, () => this.save_in_localStorage())
  }
  
  remove_task = (task_title: string) => {
    const new_list = this.state.task_list.filter(
      (arr_task: ITask) => task_title !== arr_task.title
    )
    this.setState({
      task_list: new_list
    }, () => {
      this.save_in_localStorage()
      window.location.reload()
    })
  }

  set_field = (e: any) => {
    const { name, value } = e.target
    this.setState({
      [name]: value
    })
  }

  render() {
    return (
      <ToDoContext.Provider 
        value={{
          task_list: this.state.task_list,
          add_task: this.add_task,
          edit_task: this.edit_task,
          remove_task: this.remove_task
        }}>
        <h1> Basic ToDo </h1>
        <div className="container">
          <ToDoContext.Consumer>
            {({ task_list, add_task }) => (
              <div className="content">
                <div className="title-box">
                  <Input
                    defaultValue={this.state.title}
                    name="title"
                    className="MuiInputBase-input title"
                    onChange={(e: any) => this.set_field(e)} />
                  <Button variant="outlined" onClick={add_task}>
                    Create
                  </Button>
                </div>
                <div className="task-list">
                  { task_list.map((task, i) =>
                    <TaskView a_task={task} key={i} />
                  ) }
                </div>
              </div>
            )}
          </ToDoContext.Consumer>
        </div>
      </ToDoContext.Provider>
    )
  }
}

export default App;
