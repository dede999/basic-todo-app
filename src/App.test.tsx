import App from './App'
import React from 'react'
import { render, cleanup, fireEvent } from '@testing-library/react'

const setup = (label: string) => {
  const utils = render(<App />)
  const input = utils.getByLabelText(label)
  return {
    input,
    ...utils,
  }
}

describe("App.tsx", () => {
  
  afterEach(cleanup)

  it("should be able to render the component", () => {
    const { asFragment } = render(<App />)
    expect(asFragment(<App />)).toMatchSnapshot()
  })

  it("should create a task and show it", () => {
    const { getByText, getAllByText } = render(<App />)
    const element = getByText(/create/i)
    fireEvent.click(element)
    expect(getAllByText(/delete/i)).toHaveLength(1)
  })

  it ("should be able to update some task", () => {
    const { getByText, getAllByText, getByTestId } = render(<App />)
    fireEvent.click(getByText(/create/i))
    const edit = getAllByText(/edit/i)[0]
   
    // start editing
    fireEvent.click(edit)
    fireEvent.change(getByTestId(/editing/), {
      target: { value: "My Super Task" }
    })
    fireEvent.click(getAllByText(/save/i)[0])
    // done editing
     
    expect(getByText(/^My/i)).toBeTruthy
  })

  it ("should delete a task", () => {
    const { getByText, getAllByText, getByTestId } = render(<App />)

    fireEvent.change(getByTestId(/creating/), {
      target: { value: "My Super Test" }
    })
    fireEvent.click(getByText(/create/i))
    const number_of_itens = getAllByText(/edit/i).length
    console.log(number_of_itens)

    // deleting an item
    fireEvent.click(getAllByText(/delete/i)[0])
    expect(getAllByText(/edit/i).length).toBeLessThan(number_of_itens)
  })
})