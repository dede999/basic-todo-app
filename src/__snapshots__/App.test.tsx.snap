// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.tsx should be able to render the component 1`] = `
<DocumentFragment>
  <h1>
     Basic ToDo 
  </h1>
  <div
    class="container"
  >
    <div
      class="content"
    >
      <div
        class="title-box"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-input title"
        >
          <input
            class="MuiInputBase-input MuiInput-input"
            data-testid="creating"
            name="title"
            type="text"
            value="Some task ..."
          />
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Create
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="task-list"
      />
    </div>
  </div>
</DocumentFragment>
`;
